src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -5385,6 +5385,471 @@ address generate_avx_ghash_processBlocks() {
+    __ vpxor(zmm_dVec, zmm_dVec, zmm_aVec, Assembler::AVX_512bit);
+    __ evprold(zmm_dVec, zmm_dVec, 8, Assembler::AVX_512bit);
+
+    // c += d; b ^= c; b <<<= 7 (b << 7 | scratch >>> 25)
+    __ vpaddd(zmm_cVec, zmm_cVec, zmm_dVec, Assembler::AVX_512bit);
+    __ vpxor(zmm_bVec, zmm_bVec, zmm_cVec, Assembler::AVX_512bit);
+    __ evprold(zmm_bVec, zmm_bVec, 7, Assembler::AVX_512bit);
+
+    // Before we start the next iteration, we need to perform shuffles
+    // on the b/c/d vectors to move them back to columnar organizations
+    // from their current diagonal orientation.
+    __ vpshufd(zmm_bVec, zmm_bVec, 0x93, Assembler::AVX_512bit);
+    __ vpshufd(zmm_cVec, zmm_cVec, 0x4E, Assembler::AVX_512bit);
+    __ vpshufd(zmm_dVec, zmm_dVec, 0x39, Assembler::AVX_512bit);
+
+    __ decrement(loopCounter);
+    __ jcc(Assembler::notZero, L_twoRounds);
+
+    // Add the initial state to the end register values
+    // We use a scratch register to replicate the 128-bit
+    // state data for the A/B/C/D vectors.  We will also add in
+    // the counter add mask onto zmm3 after adding in the start state.
+    __ evbroadcasti32x4(zmm_scratch, Address(state, 0), Assembler::AVX_512bit);
+    __ vpaddd(zmm_aVec, zmm_aVec, zmm_scratch, Assembler::AVX_512bit);
+    __ evbroadcasti32x4(zmm_scratch, Address(state, 16), Assembler::AVX_512bit);
+    __ vpaddd(zmm_bVec, zmm_bVec, zmm_scratch, Assembler::AVX_512bit);
+    __ evbroadcasti32x4(zmm_scratch, Address(state, 32), Assembler::AVX_512bit);
+    __ vpaddd(zmm_cVec, zmm_cVec, zmm_scratch, Assembler::AVX_512bit);
+    __ evbroadcasti32x4(zmm_scratch, Address(state, 48), Assembler::AVX_512bit);
+    __ vpaddd(zmm_dVec, zmm_dVec, zmm_scratch, Assembler::AVX_512bit);
+    __ vpaddd(zmm_dVec, zmm_dVec, ExternalAddress(StubRoutines::x86::chacha20_counter_addmask_avx512()), Assembler::AVX_512bit, rax);
+
+    // Write the ZMM state registers out to the key stream buffer
+    // Each ZMM is divided into 4 128-bit segments.  Each segment
+    // is written to memory at 64-byte displacements from one
+    // another.  The result is that all 4 blocks will be in their
+    // proper order when serialized.
+    __ evmovdqul(zmm_scratch, ExternalAddress(StubRoutines::x86::chacha20_scattermask_avx512()), Assembler::AVX_512bit, rax);
+    __ kxnorwl(writeMask, writeMask, writeMask);
+    __ evpscatterdd(Address(result, zmm_scratch, Address::times_4), writeMask, zmm_aVec, Assembler::AVX_512bit);
+    __ knotwl(writeMask, writeMask);
+    __ evpscatterdd(Address(result, zmm_scratch, Address::times_4, 16), writeMask, zmm_bVec, Assembler::AVX_512bit);
+    __ knotwl(writeMask, writeMask);
+    __ evpscatterdd(Address(result, zmm_scratch, Address::times_4, 32), writeMask, zmm_cVec, Assembler::AVX_512bit);
+    __ knotwl(writeMask, writeMask);
+    __ evpscatterdd(Address(result, zmm_scratch, Address::times_4, 48), writeMask, zmm_dVec, Assembler::AVX_512bit);
```

Comment:
Using the vextracti32x4 instead of evpscatterdd would give better performance:
    __ vextracti32x4(Address(result, 0), zmm_aVec, 0);
    __ vextracti32x4(Address(result, 64), zmm_aVec, 1);
    __ vextracti32x4(Address(result, 128), zmm_aVec, 2);
    __ vextracti32x4(Address(result, 192), zmm_aVec, 3);
    __ vextracti32x4(Address(result, 16), zmm_bVec, 0);
    __ vextracti32x4(Address(result, 80), zmm_bVec, 1);
    __ vextracti32x4(Address(result, 144), zmm_bVec, 2);
    __ vextracti32x4(Address(result, 208), zmm_bVec, 3);
    __ vextracti32x4(Address(result, 32), zmm_cVec, 0);
    __ vextracti32x4(Address(result, 96), zmm_cVec, 1);
    __ vextracti32x4(Address(result, 160), zmm_cVec, 2);
    __ vextracti32x4(Address(result, 224), zmm_cVec, 3);
    __ vextracti32x4(Address(result, 48), zmm_dVec, 0);
    __ vextracti32x4(Address(result, 112), zmm_dVec, 1);
    __ vextracti32x4(Address(result, 176), zmm_dVec, 2);
    __ vextracti32x4(Address(result, 240), zmm_dVec, 3);

Reply:
I have been wondering about this approach for a while now, since I did something similar for the AVX2 version.  I had assumed that using evpscatterdd used less instructions and therefore would be more efficient, but I'm more than happy to move to the vextracti32x4 approach.  I'll be eager to see how it impacts performance along with the increased storage of intermediate data on additional XMMRegister objects.

Reply:
The changes you recommended yielded about a 10-15% performance improvement on the system I was using for benchmarks.  Thanks for the suggestions!